### ISSUE TITLE
Variant ID is overridden by external ID

### ISSUE THREAD

**Issue by svengau** [ynnoj/gatsby-source-printful#15]
Hi @ynnoj, 

when the plugin retrieves data, the variant ID is overridden by external ID here: https://github.com/ynnoj/gatsby-source-printful/blob/2585e522c09ae944563b8423ea03ae760efe86c4/gatsby-source-printful/gatsby-node.js#L100

But we need the variant Id to create the order (or also retrieve the variant using the API). If it's a breaking change, maybe we can rename id to variant_id ? 

What do you think ?
And thanks for this useful plugin :)

**Comment by ynnoj** [comment]
@svengau As I understand, you just want the `variant_id` returned by Printful available on the variant node?

If so, I just need to not destructure it here:
https://github.com/ynnoj/gatsby-source-printful/blob/2585e522c09ae944563b8423ea03ae760efe86c4/gatsby-source-printful/gatsby-node.js#L111

**Comment by svengau** [comment]
I didn't see a field "variant_id" exists, I thought about "id" :)

According the doc, https://www.printful.com/docs/orders#actionGet, we have:

```
id	integer	Line item ID
external_id	string	Line item ID from the external system
variant_id	integer	Variant ID of the item ordered. See Products API
sync_variant_id	integer	Sync variant ID of the item ordered.
external_variant_id	string	External variant ID of the item ordered.
warehouse_product_variant_id string Warehousing product variant ID of the item ordered. See Warehouse Products API
```

https://www.printful.com/docs/products#getSingleSyncVariant


https://www.printful.com/docs/products#getSingleSyncVariant

**Comment by svengau** [comment]
@ynnoj I propose this breaking PR also. 

Same philosophy than issue https://github.com/ynnoj/gatsby-source-printful/issues/14, the best option is to not change / remap the data coming from Prinful.

Maybe the best is to push a major version, and add a breaking changes in README.md. WDYT ?

**Comment by LuigiClaudio** [comment]
> I didn't see a field "variant_id" exists, I thought about "id" :)
> 
> According the doc, https://www.printful.com/docs/orders#actionGet, we have:
> 
> ```
> id	integer	Line item ID
> external_id	string	Line item ID from the external system
> variant_id	integer	Variant ID of the item ordered. See Products API
> sync_variant_id	integer	Sync variant ID of the item ordered.
> external_variant_id	string	External variant ID of the item ordered.
> warehouse_product_variant_id string Warehousing product variant ID of the item ordered. See Warehouse Products API
> ```
> 
> But when I play with the getSingleSyncVariant API:
> 
> https://www.printful.com/docs/products#getSingleSyncVariant
> 
> it only works with the ID.
> I think the best is still to send back the exact response from printful.
> 
> WDYT ?
Did you try with prefixing it with the @ symbol as mentioned there?  It works.
`Sync Variant ID (integer) or External ID (if prefixed with @)`

**Comment by LuigiClaudio** [comment]
@svengau Did you try with prefixing it with the @ symbol as mentioned there?  It works.
`Sync Variant ID (integer) or External ID (if prefixed with @)`

**Comment by ynnoj** [comment]
You can try this out on [`2.0.0-beta.1`](https://github.com/ynnoj/gatsby-source-printful/releases/tag/v2.0.0-beta.1).

**Comment by svengau** [comment]
Hi @ynnoj, the beta version works fine for me.

**Comment by LuigiClaudio** [comment]
this stopped working for me: getting following error.

`"gatsby-source-printful" threw an error while running the sourceNodes lifecycle:

products.map(...).flat is not a function

  63 |       variants.map(({ variant_id }) => variant_id)
  64 |     )
> 65 |     .flat()
     |      ^
  66 |
  67 |   const uniqueCatalogVariantIds = catalogVariantIds.reduce(
  68 |     (unique, item) => (unique.includes(item) ? unique : [...unique, item]),

File: ..\node_modules\gatsby-source-printful\gatsby-node.js:65:6`

**Comment by LuigiClaudio** [comment]
After replacing the `flat()` method with `reduce()` on my local node modules, the Printful node are created just fine (awesome!) but I see this warning in the terminal: 
```
warn There are conflicting field types in your data.

If you have explicitly defined a type for those fields, you can safely ignore this warning message.
Otherwise, Gatsby will omit those fields from the GraphQL schema.

If you know all field types in advance, the best strategy is to explicitly define them with the `createTypes` action, and skip inference with the `@dontInfer` directive.
See https://www.gatsbyjs.org/docs/actions/#createTypes
PrintfulCatalogProduct.options.values:
 - type: array
   value: [ ..., 'With pocket', ... ]
 - type: object
   value: { white: 'White', black: 'Black' }
```

**Comment by ynnoj** [comment]
> This stopped working for me. Getting following error:
> 
> ```
> "gatsby-source-printful" threw an error while running the sourceNodes lifecycle:
> 
> products.map(...).flat is not a function
> 
>   63 |       variants.map(({ variant_id }) => variant_id)
>   64 |     )
> > 65 |     .flat()
>      |      ^
>   66 |
>   67 |   const uniqueCatalogVariantIds = catalogVariantIds.reduce(
>   68 |     (unique, item) => (unique.includes(item) ? unique : [...unique, item]),
> 
> File: ..\node_modules\gatsby-source-printful\gatsby-node.js:65:6
> ```

What version of Node are you using?

**Comment by ynnoj** [comment]
> After replacing the `flat()` method with `reduce()` on my local node modules, the Printful nodes are created just fine (awesome!) but I see this warning in the terminal:
> 
> ```
> warn There are conflicting field types in your data.
> 
> If you have explicitly defined a type for those fields, you can safely ignore this warning message.
> Otherwise, Gatsby will omit those fields from the GraphQL schema.
> 
> If you know all field types in advance, the best strategy is to explicitly define them with the `createTypes` action, and skip inference with the `@dontInfer` directive.
> See https://www.gatsbyjs.org/docs/actions/#createTypes
> PrintfulCatalogProduct.options.values:
>  - type: array
>    value: [ ..., 'With pocket', ... ]
>  - type: object
>    value: { white: 'White', black: 'Black' }
> ```

Uh, inconsistent types from the Printful API. Great 😑

**Comment by LuigiClaudio** [comment]
> > This stopped working for me. Getting following error:
> > ```
> > "gatsby-source-printful" threw an error while running the sourceNodes lifecycle:
> > 
> > products.map(...).flat is not a function
> > 
> >   63 |       variants.map(({ variant_id }) => variant_id)
> >   64 |     )
> > > 65 |     .flat()
> >      |      ^
> >   66 |
> >   67 |   const uniqueCatalogVariantIds = catalogVariantIds.reduce(
> >   68 |     (unique, item) => (unique.includes(item) ? unique : [...unique, item]),
> > 
> > File: ..\node_modules\gatsby-source-printful\gatsby-node.js:65:6
> > ```
> 
> @LuigiClaudio What version of Node are you using?

`10.18.0` - I gave it a go with `11.1.0` - same result.

**Comment by LuigiClaudio** [comment]
> > > This stopped working for me. Getting following error:
> > > ```
> > > "gatsby-source-printful" threw an error while running the sourceNodes lifecycle:
> > > 
> > > products.map(...).flat is not a function
> > > 
> > >   63 |       variants.map(({ variant_id }) => variant_id)
> > >   64 |     )
> > > > 65 |     .flat()
> > >      |      ^
> > >   66 |
> > >   67 |   const uniqueCatalogVariantIds = catalogVariantIds.reduce(
> > >   68 |     (unique, item) => (unique.includes(item) ? unique : [...unique, item]),
> > > 
> > > File: ..\node_modules\gatsby-source-printful\gatsby-node.js:65:6
> > > ```
> > 
> > 
> > @LuigiClaudio What version of Node are you using?
> 
> `10.18.0` - I gave it a go with `11.1.0` - same result.

@ynnoj think that neither of those versions make the cut for using the `flat()` method. The method is available from version `11.15.0`
https://node.green/#ES2019-features-Array-prototype--flat--flatMap--Array-prototype-flat

**Comment by ynnoj** [comment]
@LuigiClaudio Sweet! I’d recommend using at least version 12 (preferably 14) inline with the LTS. 

https://nodejs.org/en/about/releases/
